fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+        with multiprocessing.dummy.Pool(num_workers) as pool:
+            for is_success, url, error_type in pool.imap_unordered(
+                _do_download, tasks
+            ):
+                if is_success:
+                    if len(downloaded) < max_videos:
+                        downloaded.append(url)
+                        pb.update()
+                    else:
+                        return downloaded, errors
+                else:
+                    errors[error_type].append(url)
+    return downloaded, errors
+
+
+def _build_ydl_opts(ext, video_path, clip_segment, download_dir):
+    ytdl_logger = logging.getLogger("ytdl-ignore")
+    ytdl_logger.disabled = True
+    ydl_opts = {
+        "age_limit": 99,
+        "logger": ytdl_logger,
+    }
+
+    if video_path:
+        output_path = video_path
+        _ext = os.path.splitext(os.path.basename(video_path))[1]
+        _ext = _ext.lstrip(".")
+
+    else:
+        if ext is not None:
+            _ext = ext.lstrip(".")
+            path_ext = _ext
+        else:
+            _ext = None
+            path_ext = "%(ext)s"
+        path_name = "%(title)s"
+        output_path = os.path.join(
+            download_dir, "%s.%s" % (path_name, path_ext)
+        )
+
+    if _ext:
+        ydl_opts["format"] = _ext
+
+    ydl_opts["outtmpl"] = output_path
+
+    return ydl_opts, output_path
+
+
+def _do_download(args):
+    url, video_path, clip_segment, download_dir, ext, buffer_seconds = args
+    ydl_opts, output_path = _build_ydl_opts(
+        ext, video_path, clip_segment, download_dir
+    )
+
+    try:
+        if clip_segment in [None, [None, None], [0, None]]:
+            with youtube_dl.YoutubeDL(ydl_opts) as ydl:
+                ydl.download([url])
+        else:
+            with youtube_dl.YoutubeDL(ydl_opts) as ydl:
+                info_dict = ydl.extract_info(url, download=False)
+                _url = info_dict.get("url", None)
+                if _url is None:
+                    _url = info_dict["requested_formats"][0]["url"]
+
+                _id = info_dict.get("id", None)
+                _title = info_dict.get("title", _id)
+                _ext = info_dict.get("ext", ext)
+
+                output_path = output_path.replace(
+                    "%(title)s", _title.replace("/", "")
+                )
+                output_path = output_path.replace("%(ext)s", _ext)
+
+                _download_ffmpeg(
+                    _url,
+                    output_path,
+                    clip_segment,
+                    buffer_seconds=buffer_seconds,
+                )
+
+        return True, url, None
+
+    except Exception as e:
+        if isinstance(e, youtube_dl.utils.DownloadError):
+            # pylint: disable=no-member
+            return False, url, str(e.exc_info[1])
+        return False, url, "other"
+
+
+def _download_ffmpeg(url, output_path, clip_segment, buffer_seconds=5):
```

Comment:
This is nearly a re-implementation of what you can achieve with:

```py
import fiftyone.utils.video as fouv

fouv.extract_clip(url, output_path, timestamps=clip_segment)
```

It'd be nice to avoid code duplication. If that other method needs some additional parameters to deal with accuracy or other issues, then please add them there :)